File: py/all_until_nth_prime.py
# Willans'_formula - all_until_nth_prime.py
# Copyright (c) 2025 epsill0n epsill0n
# Non-commercial use only. Redistribution or modification prohibited without permission.
# See LICENSE.txt for full terms.

import sys
import time
from WF import Willans
nthPrime_seq = Willans.seq.nthPrime
nthPrime_conc = Willans.conc.nthPrime
import mpmath

if __name__=='__main__':

    if (len(sys.argv)!=4):
        print(f'Usage: [{sys.argv[0]}] [n] [1/0 for concurrency] [1/0 for accumulation]')
        exit(-1)
    
    n = int(sys.argv[1])
    isConcurrent = int(sys.argv[2])
    isCumulative = int(sys.argv[3])

    print(f'Using mpmath precision of {mpmath.mp.dps} decimals.')

    Willans.init_funcs(isConcurrent, isCumulative)

    if isConcurrent:
        print("Using concurrency.")
        nthPrime = nthPrime_conc
    else:
        print("Not using concurrency.")
        nthPrime = nthPrime_seq

    if isCumulative:
        print("Using accumulation.")
    else:
        print("Not using accumulation.")
    
    primes = [_ for _ in range(n)]
    start = time.time()
    for i in range(1, n+1):
        primes[i-1] = nthPrime(i)
    end = time.time()

    for i in range (1, n+1):
        print(f'{primes[i-1]}', end=' ')
    
    print(f'\nTime: {(end - start) * 1e3}ms')

File: cpp/all_until_nth_prime.cpp
// Willans'_formula - all_until_nth_prime.cpp
// Copyright (c) 2025 epsill0n epsill0n
// Non-commercial use only. Redistribution or modification prohibited without permission.
// See LICENSE.txt for full terms.

#include <iostream>
#   define FAST_IO
#include <iomanip>

#include <cstdlib>

#   define GCC_O3_OPTIM
#include "WF_impl.hpp"

#include <chrono>


__attribute__((constructor)) void init() {
#ifdef FAST_IO
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(0);
#endif
    WF::init_mpfr(MPFR_PRECISION);
}


int main(int argc, char** argv) {
    if (argc!=3) {
        std::cout << "Usage: [" << argv[0] << "] [n] [1/0 for concurrency]\n" << std::flush;
        return -1;
    }

    std::cout << "Using MPFR precision of " << MPFR_PRECISION << " bits.\n";

    int n = std::atoi(argv[1]);
    mpz_class prime[n];
    std::chrono::system_clock::time_point begin, end;
    std::chrono::duration<double, std::milli> elapsed;
    if (std::atoi(argv[2]) == 0) {
        std::cout << "Not using concurrency.\n" << std::flush;

        begin = std::chrono::high_resolution_clock::now();
        for (int i = 0; i < n; ++i)
            prime[i] = WF::seq::nthPrime(i+1);

        end = std::chrono::high_resolution_clock::now();

    } else {
        std::cout << "Using concurrency.\n" << std::flush;
        begin = std::chrono::high_resolution_clock::now();
        for (int i = 0; i < n; ++i)
            prime[i] = WF::conc::nthPrime(i+1);
        end = std::chrono::high_resolution_clock::now();
    }
    
    elapsed = end - begin;
    for (int i = 0; i < n; ++i)
        std::cout << prime[i] << " ";
    std::cout << "\nTime: " << std::fixed << std::setprecision(10) <<  (elapsed.count()) << "ms\n";
    
    std::cout << std::flush;
}